nifi-nar-bundles/nifi-druid-bundle/nifi-druid-processors/src/main/java/org/apache/nifi/processors/PutDruid.java

```diff
@@ -0,0 +1,206 @@
+
+@SideEffectFree
+@Tags({"Druid","Timeseries","OLAP","ingest"})
+@CapabilityDescription("Sends events to Apache Druid for Indexing. "
+						+ "Leverages Druid Tranquility Controller service."
+						+ "Incoming flow files are expected to contain 1 or many JSON objects, one JSON object per line")
+public class PutDruid extends AbstractSessionFactoryProcessor {
+
+    private List<PropertyDescriptor> properties;
+    private Set<Relationship> relationships;
+    private final Map<Object,String> messageStatus = new HashMap<Object,String>();
+
+    public static final PropertyDescriptor DRUID_TRANQUILITY_SERVICE = new PropertyDescriptor.Builder()
+            .name("druid_tranquility_service")
+            .description("Tranquility Service to use for sending events to Druid")
+            .required(true)
+            .identifiesControllerService(DruidTranquilityService.class)
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("SUCCESS")
+            .description("Succes relationship")
+            .build();
+
+    public static final Relationship REL_FAIL = new Relationship.Builder()
+            .name("FAIL")
+            .description("FlowFiles are routed to this relationship when they cannot be parsed")
+            .build();
+
+    public static final Relationship REL_DROPPED = new Relationship.Builder()
+            .name("DROPPED")
+            .description("FlowFiles are routed to this relationship when they are outside of the configured time window, timestamp format is invalid, ect...")
+            .build();
+
+    public void init(final ProcessorInitializationContext context){
+        List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(DRUID_TRANQUILITY_SERVICE);
+        this.properties = Collections.unmodifiableList(properties);
+
+        Set<Relationship> relationships = new HashSet<Relationship>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_DROPPED);
+        relationships.add(REL_FAIL);
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships(){
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+    
+    //Method breaks down incoming flow file and sends it to Druid Indexing servic
+    private void processFlowFile(ProcessContext context, ProcessSession session){
+    	 //Get handle on Druid Tranquility session
+    	 DruidTranquilityService tranquilityController = context.getProperty(DRUID_TRANQUILITY_SERVICE)
+                 .asControllerService(DruidTranquilityService.class);
+         Tranquilizer<Map<String,Object>> tranquilizer = tranquilityController.getTranquilizer();
+         
+         final FlowFile flowFile = session.get();
+         if (flowFile == null || flowFile.getSize() == 0) {
```

Comment:
if you've pulled a flowfile of size 0 and just return you'll not have handled the size 0 objects.  I'd recommend having the size check in another block and terminating it or routing it to empty objects or something like that.  

also consider doing this check before getting the controller service handle.

Reply:
I was following the example from the AbstractHBase processor:

https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/AbstractPutHBase.java#L137

Could you clarify what you mean by terminating it?